incapsula/resource_abp_websites.go

```diff
@@ -0,0 +1,403 @@
+)
+
+type AbpTerraformAccount struct {
+	AutoPublish   bool                       `json:"auto_publish"`
+	WebsiteGroups []AbpTerraformWebsiteGroup `json:"website_groups"`
+}
+
+type AbpTerraformWebsiteGroup struct {
+	Id       *string `json:"id"`
+	NameId   *string
+	Name     string                `json:"name"`
+	Websites []AbpTerraformWebsite `json:"websites"`
+}
+
+type AbpTerraformWebsite struct {
+	Id               *string `json:"id"`
+	WebsiteId        int     `json:"website_id"`
+	EnableMitigation bool    `json:"enable_mitigation"`
+}
+
+func (a *AbpTerraformWebsiteGroup) UniqueId() string {
+	if a.NameId == nil {
+		return a.Name
+	}
+	return *a.NameId
+}
+
+func resourceAbpWebsites() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: resourceAbpWebsitesCreate,
+		ReadContext:   resourceAbpWebsitesRead,
+		UpdateContext: resourceAbpWebsitesUpdate,
+		DeleteContext: resourceAbpWebsitesDelete,
+
+		Importer: &schema.ResourceImporter{
+			StateContext: func(ctx context.Context, d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
+				accountIdStr := d.Id()
+				accountId, err := strconv.Atoi(accountIdStr)
+
+				if err != nil {
+					return nil, fmt.Errorf("Expected an account Id which must be an integer: %s", err.Error())
+				}
+
+				d.Set("account_id", accountId)
+				d.SetId(accountIdStr)
+
+				client := meta.(*Client)
+				abpWebsites, diags := client.ReadAbpWebsites(accountId)
+
+				for _, diagnostic := range diags {
+					if diagnostic.Severity == diag.Error {
+						return nil, fmt.Errorf("%s", diagnostic.Detail)
+					}
+				}
+
+				setUniqueNameIds(abpWebsites)
+
+				serializeAccount(d, *abpWebsites)
+
+				log.Printf("[DEBUG] Import ABP websites for account id %d", accountId)
+
+				return []*schema.ResourceData{d}, nil
+			},
+		},
+
+		Description: "Provides an Incapsula ABP (Advanced Bot Protection) website resource. Allows you to enable and configure ABP for given websites.",
+
+		Schema: map[string]*schema.Schema{
+			"account_id": {
+				Description: "The account these websites belongs to.",
+				Type:        schema.TypeInt,
+				Required:    true,
+			},
+			"auto_publish": {
+				Description: "Whether to publish the changes automatically. Changes don't take take effect until they have been published.",
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Default:     false,
+			},
+			"last_publish": {
+				Description: "When the last publish was done for this terraform resource. Changes are published when `auto_publish` is true and the terraform config is applied.",
+				Type:        schema.TypeString,
+				Computed:    true,
+			},
+			"website_group": {
+				Description: "List of website groups which are associated to ABP. Website groups are matched in a top-down fashion. If a more specific website group should take precedence over over a wild card entry then that should be higher in the list.",
+				Type:        schema.TypeList,
```

Comment:
I looked at using `TypeSet` instead (as discussed) but since the order that the website groups (and websites) are defined in matters I don't believe that works and it has to be a list.

Reply:
We need to document that it has semantic impact to change the order of website {groups} in the tf file..

Reply:
It is documented here? Do you want the wording to change?